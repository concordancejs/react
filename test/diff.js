import test from 'ava'
import concordance from 'concordance'
import React from 'react'
import renderer from 'react-test-renderer'

import plugin from '..'
import HelloMessage from './fixtures/react/HelloMessage'

const plugins = [plugin]

const diff = (t, getSame, getOther, options) => {
  t.snapshot(concordance.diff(getSame(), getOther(), Object.assign({plugins}, options)))
}
diff.title = prefix => `diffs ${prefix}`
const diffRendered = (t, getSame, getOther, options) => {
  const render = value => renderer.create(value).toJSON()
  t.snapshot(concordance.diff(render(getSame()), render(getOther()), Object.assign({plugins}, options)))
}
diffRendered.title = prefix => `diffs rendered ${prefix}`

const macros = [diff, diffRendered]

test('property differences between react elements', macros,
  () => <HelloMessage name='John' />,
  () => <HelloMessage name='Olivia' />)

test('react elements', macros,
  () => <strong>arm</strong>,
  () => <em>arm</em>)

test('fragments', macros,
  () => <React.Fragment><HelloMessage name='John' /></React.Fragment>,
  () => <React.Fragment><HelloMessage name='Olivia' /></React.Fragment>)

test('object properties', macros,
  () => React.createElement('Foo', {object: {baz: 'thud'}}),
  () => React.createElement('Foo', {object: {baz: 'qux'}}))

test('array values in object properties', macros,
  () => React.createElement('Foo', {object: {baz: ['thud']}}),
  () => React.createElement('Foo', {object: {baz: ['qux']}}))

test('array values in object properties in children', macros,
  () => {
    return React.createElement('Foo', null, React.createElement('Bar', {object: {baz: ['thud']}}))
  },
  () => {
    return React.createElement('Foo', null, React.createElement('Bar', {object: {baz: ['qux']}}))
  })

test('same name, different component function', t => {
  const actual = <HelloMessage name='John' />
  class Faux extends React.Component {}
  Faux.displayName = 'HelloMessage'
  const expected = <Faux name='John' />
  t.snapshot(concordance.diff(actual, expected, {plugins}))
})

test('multiline string properties', macros,
  () => React.createElement('Foo', {multiline: 'foo\nbar'}),
  () => React.createElement('Foo', {multiline: 'foo\nbaz'}))

test('string and number children', macros,
  () => React.createElement('div', null, 'foo', 'bar', 42),
  () => React.createElement('div', null, 'foo bar', 39))

test('both have properties and children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Foo', {baz: 'thud'}, React.createElement('Baz')))

test('both have properties and children, different names', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Qux', {baz: 'thud'}, React.createElement('Baz')))

test('both have properties, first has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Foo', {baz: 'thud'}))

test('both have properties, second has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}),
  () => React.createElement('Foo', {baz: 'thud'}, React.createElement('Baz')))

test('both have properties, neither has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}),
  () => React.createElement('Foo', {baz: 'thud'}))

test('first has properties, both have children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Foo', null, React.createElement('Baz')))

test('first has properties, both have children, different names', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Qux', null, React.createElement('Baz')))

test('first has properties, first has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}, React.createElement('Bar')),
  () => React.createElement('Foo'))

test('first has properties, second has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}),
  () => React.createElement('Foo', null, React.createElement('Baz')))

test('first has properties, neither has children, same name', macros,
  () => React.createElement('Foo', {foo: 'bar'}),
  () => React.createElement('Foo'))

test('second has properties, both have children, same name', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Foo', {baz: 'thud'}, React.createElement('Baz')))

test('second has properties, both have children, different names', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Qux', {baz: 'thud'}, React.createElement('Baz')))

test('second has properties, first has children, same name', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Foo', {baz: 'thud'}))

test('second has properties, second has children, same name', macros,
  () => React.createElement('Foo'),
  () => React.createElement('Foo', {baz: 'thud'}, React.createElement('Baz')))

test('second has properties, neither has children, same name', macros,
  () => React.createElement('Foo'),
  () => React.createElement('Foo', {baz: 'thud'}))

test('neither have properties, both have children, same name', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Foo', null, React.createElement('Baz')))

test('neither have properties, both have children, different names', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Qux', null, React.createElement('Baz')))

test('neither have properties, first has children, same name', macros,
  () => React.createElement('Foo', null, React.createElement('Bar')),
  () => React.createElement('Foo'))

test('neither have properties, second has children, same name', macros,
  () => React.createElement('Foo'),
  () => React.createElement('Foo', null, React.createElement('Baz')))

test('neither have properties, neither has children, same name', macros,
  () => React.createElement('Foo'),
  () => React.createElement('Foo'))

test('max depth', macros,
  () => (
    <div>
      <div>
        <div>
          Hello
        </div>
        <div id='foo'>Hello</div>
        <br id='bar'/>
        <br/>
        <section>
          <div>
            Foo
          </div>
        </section>
      </div>
    </div>
  ),
  () => (
    <div>
      <div>
        <div>
          Hello
        </div>
        <div id='foo'>Hello</div>
        <br id='bar'/>
        <br/>
        <section>
          <div>
            Bar
          </div>
        </section>
      </div>
    </div>
  ),
  {maxDepth: 2})
