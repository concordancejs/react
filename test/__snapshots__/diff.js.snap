// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffs array values in object properties 1`] = `
"  <Foo
    object={
      Object {
        baz: Array [
-         'thud',
+         'qux',
        ],
      }
    }
  />"
`;

exports[`diffs array values in object properties in children 1`] = `
"  <Foo>
    <Bar
      object={
        Object {
          baz: Array [
-           'thud',
+           'qux',
          ],
        }
      }
    />
  </Foo>"
`;

exports[`diffs both have properties and children, different names 1`] = `
"- <Foo
-   foo=\\"bar\\"
+ <Qux
+   baz=\\"thud\\"
  >
-   <Bar />
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs both have properties and children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
+   baz=\\"thud\\"
  >
-   <Bar />
+   <Baz />
  </Foo>"
`;

exports[`diffs both have properties, first has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
- </Foo>
+   baz=\\"thud\\"
+ />"
`;

exports[`diffs both have properties, neither has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
+   baz=\\"thud\\"
  />"
`;

exports[`diffs both have properties, second has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
- />
+   baz=\\"thud\\"
+ >
+   <Baz />
+ </Foo>"
`;

exports[`diffs first has properties, both have children, different names 1`] = `
"- <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
+ <Qux>
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs first has properties, both have children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
+ <Foo>
+   <Baz />
  </Foo>"
`;

exports[`diffs first has properties, first has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
-   <Bar />
- </Foo>
+ <Foo />"
`;

exports[`diffs first has properties, neither has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
+ <Foo />"
`;

exports[`diffs first has properties, second has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
+ <Foo>
+   <Baz />
+ </Foo>"
`;

exports[`diffs max depth 1`] = `
"  <div>
    <div>
      <div>…</div>
      <div…>…</div>
      <br… />
      <br />
      <section>
        <div>
-         Foo
+         Bar
        </div>
      </section>
    </div>
  </div>"
`;

exports[`diffs multiline string properties 1`] = `
"  <Foo
    multiline={
      \`foo␊
-     bar\`
+     baz\`
    }
  />"
`;

exports[`diffs neither have properties, both have children, different names 1`] = `
"- <Foo>
-   <Bar />
+ <Qux>
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs neither have properties, both have children, same name 1`] = `
"  <Foo>
-   <Bar />
+   <Baz />
  </Foo>"
`;

exports[`diffs neither have properties, first has children, same name 1`] = `
"- <Foo>
-   <Bar />
- </Foo>
+ <Foo />"
`;

exports[`diffs neither have properties, neither has children, same name 1`] = `"  <Foo />"`;

exports[`diffs neither have properties, second has children, same name 1`] = `
"- <Foo />
+ <Foo>
+   <Baz />
+ </Foo>"
`;

exports[`diffs object properties 1`] = `
"  <Foo
    object={
      Object {
-       baz: 'thud',
+       baz: 'qux',
      }
    }
  />"
`;

exports[`diffs property differences between react elements 1`] = `
"  <HelloMessage⍟
-   name=\\"John\\"
+   name=\\"Olivia\\"
  />"
`;

exports[`diffs react elements 1`] = `
"- <strong>
+ <em>
    arm
- </strong>
+ </em>"
`;

exports[`diffs rendered array values in object properties 1`] = `
"  <Foo
    object={
      Object {
        baz: Array [
-         'thud',
+         'qux',
        ],
      }
    }
  />"
`;

exports[`diffs rendered array values in object properties in children 1`] = `
"  <Foo>
    <Bar
      object={
        Object {
          baz: Array [
-           'thud',
+           'qux',
          ],
        }
      }
    />
  </Foo>"
`;

exports[`diffs rendered both have properties and children, different names 1`] = `
"- <Foo
-   foo=\\"bar\\"
+ <Qux
+   baz=\\"thud\\"
  >
-   <Bar />
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs rendered both have properties and children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
+   baz=\\"thud\\"
  >
-   <Bar />
+   <Baz />
  </Foo>"
`;

exports[`diffs rendered both have properties, first has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
- </Foo>
+   baz=\\"thud\\"
+ />"
`;

exports[`diffs rendered both have properties, neither has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
+   baz=\\"thud\\"
  />"
`;

exports[`diffs rendered both have properties, second has children, same name 1`] = `
"  <Foo
-   foo=\\"bar\\"
- />
+   baz=\\"thud\\"
+ >
+   <Baz />
+ </Foo>"
`;

exports[`diffs rendered first has properties, both have children, different names 1`] = `
"- <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
+ <Qux>
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs rendered first has properties, both have children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- >
-   <Bar />
+ <Foo>
+   <Baz />
  </Foo>"
`;

exports[`diffs rendered first has properties, first has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
-   <Bar />
- </Foo>
+ <Foo />"
`;

exports[`diffs rendered first has properties, neither has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
+ <Foo />"
`;

exports[`diffs rendered first has properties, second has children, same name 1`] = `
"- <Foo
-   foo=\\"bar\\"
- />
+ <Foo>
+   <Baz />
+ </Foo>"
`;

exports[`diffs rendered max depth 1`] = `
"  <div>
    <div>
      <div>…</div>
      <div…>…</div>
      <br… />
      <br />
      <section>
        <div>
-         Foo
+         Bar
        </div>
      </section>
    </div>
  </div>"
`;

exports[`diffs rendered multiline string properties 1`] = `
"  <Foo
    multiline={
      \`foo␊
-     bar\`
+     baz\`
    }
  />"
`;

exports[`diffs rendered neither have properties, both have children, different names 1`] = `
"- <Foo>
-   <Bar />
+ <Qux>
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs rendered neither have properties, both have children, same name 1`] = `
"  <Foo>
-   <Bar />
+   <Baz />
  </Foo>"
`;

exports[`diffs rendered neither have properties, first has children, same name 1`] = `
"- <Foo>
-   <Bar />
- </Foo>
+ <Foo />"
`;

exports[`diffs rendered neither have properties, neither has children, same name 1`] = `"  <Foo />"`;

exports[`diffs rendered neither have properties, second has children, same name 1`] = `
"- <Foo />
+ <Foo>
+   <Baz />
+ </Foo>"
`;

exports[`diffs rendered object properties 1`] = `
"  <Foo
    object={
      Object {
-       baz: 'thud',
+       baz: 'qux',
      }
    }
  />"
`;

exports[`diffs rendered property differences between react elements 1`] = `
"  <div>
    Hello 
    <mark>
-     John
+     Olivia
    </mark>
  </div>"
`;

exports[`diffs rendered react elements 1`] = `
"- <strong>
+ <em>
    arm
- </strong>
+ </em>"
`;

exports[`diffs rendered second has properties, both have children, different names 1`] = `
"- <Foo>
-   <Bar />
+ <Qux
+   baz=\\"thud\\"
+ >
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs rendered second has properties, both have children, same name 1`] = `
"- <Foo>
-   <Bar />
+ <Foo
+   baz=\\"thud\\"
+ >
+   <Baz />
  </Foo>"
`;

exports[`diffs rendered second has properties, first has children, same name 1`] = `
"- <Foo>
-   <Bar />
- </Foo>
+ <Foo
+   baz=\\"thud\\"
+ />"
`;

exports[`diffs rendered second has properties, neither has children, same name 1`] = `
"- <Foo />
+ <Foo
+   baz=\\"thud\\"
+  />"
`;

exports[`diffs rendered second has properties, second has children, same name 1`] = `
"- <Foo />
+ <Foo
+   baz=\\"thud\\"
+ >
+   <Baz />
+ </Foo>"
`;

exports[`diffs rendered string and number children 1`] = `
"  <div>
-   foo
+   foo bar
-   bar
-   {42}
+   {39}
  </div>"
`;

exports[`diffs second has properties, both have children, different names 1`] = `
"- <Foo>
-   <Bar />
+ <Qux
+   baz=\\"thud\\"
+ >
+   <Baz />
- </Foo>
+ </Qux>"
`;

exports[`diffs second has properties, both have children, same name 1`] = `
"- <Foo>
-   <Bar />
+ <Foo
+   baz=\\"thud\\"
+ >
+   <Baz />
  </Foo>"
`;

exports[`diffs second has properties, first has children, same name 1`] = `
"- <Foo>
-   <Bar />
- </Foo>
+ <Foo
+   baz=\\"thud\\"
+ />"
`;

exports[`diffs second has properties, neither has children, same name 1`] = `
"- <Foo />
+ <Foo
+   baz=\\"thud\\"
+  />"
`;

exports[`diffs second has properties, second has children, same name 1`] = `
"- <Foo />
+ <Foo
+   baz=\\"thud\\"
+ >
+   <Baz />
+ </Foo>"
`;

exports[`diffs string and number children 1`] = `
"  <div>
-   foo
+   foo bar
-   bar
-   {42}
+   {39}
  </div>"
`;

exports[`same name, different component function 1`] = `
"- <HelloMessage⍟
+ <HelloMessage⍟
    name=\\"John\\"
  />"
`;
